### PR TITLE
Global ratelimiter: improve low-weight low-usage scenarios

### PR DESCRIPTION
**Pull request by Groxx**
Part 1 of 2 to address some relatively-expected "infrequent usage behaves sub-optimally" issues.

In this one, I'm tackling an issue where there are some high-usage limiting-hosts, and some occasional-and-low-usage hosts, and e.g. less than half of the target RPS is being used at any time.

In ^ this scenario it's possible for the low-usage hosts to be given a weight so low that they block _most_ of the requests they receive, especially if those requests come in small bursts (which is quite common) rather than evenly distributed.  This is partly because their weight is low enough to only allow a couple RPS, and partly because our limiters currently set their burst size == limit, so they also do not allow bursts that would otherwise be within-budget.

This _cannot_ be addressed perfectly with an async system like this (we can't perfectly predict caller behavior), but there are some easy partial improvements worth building:

1. (this PR) Allow low-weight hosts to over-commit and accept more than their fair share of RPS, if there is free RPS in the cluster.
   - For _consistently_ low-RPS hosts, this should be fair and a moderate improvement, and it performs no worse than a non-global algorithm if the limit is not usually fully used.
2. Raise the burst size, so intermittent callers can accumulate a small burst budget.
   - Currently this matches the RPS (min 1), which means e.g. a totally innocuous 2-request-event every 10 seconds will likely reject 50%.  Obviously that's not great.
   - "Fair" values here are fuzzier, and will be figured out in a later PR.

In this PR, I "boost" RPSes that are below the fallback values, in a way that converges towards the target RPS as usage increases.  This allows some over-committing, but we already allow that, and the fallback values are assumed to be "safe enough": they'd be used if the caller just waited a bit longer between calls -> the global limiter was garbage-collected.

If global usage increases, the boost-amount will shrink, and the over-commit will shrink as well.  So while this allows spiky requests through at a somewhat higher rate than desired, heavy usage should still converge near the target RPS rather than e.g. 2x the target.

---

A lot of lines have changed, but this PR boils down to:
- change the IDL to return more data (which was already tracked) as a new type
  - this will break the current system, causing the global limiters to use their fallback
  - ^ this is currently intentional, to serve as a test of that fallback, but should not generally be done.  normally it should handle both types, or shift to a plugin system and just have two stacks in use, switched by type.
- change the `algorithm` portion to return a struct rather than 2 values, to make this additional data easier to pass around
- boost RPS when updating the ratelimiters, in the `collection`
- a lot of minor changes to handle the new type, without changing behavior

---

Visually, for a hopefully more intuitive grasp of the issue, this is how weights/RPS currently get distributed among a number of hosts:
ratelimiter - orig

Low weights get a low ratio of the allowed RPS, so they are also only allowed to _grow_ a little, while high-weight hosts are allowed to grow a lot.  If we assume traffic arrives at frontends roughly randomly, this is fine (if they grow, they will all grow by the same %), but in practice we see a lot of "constant high use" to some hosts, and "occasional low use" to others, e.g. due to different services performing different call patterns and being routed differently.

Mathematically this makes sense... but it's relatively easy for something that only sends a couple requests every 10 seconds or so to eventually be allocated ~1 RPS because its weight is legitimately that low.  But then nearly all of those low-RPS requests are rejected despite there (possibly) being more than enough free-RPS in the cluster.

This PR essentially changes the behavior to look like this:
ratelimiter - changed

If there is unused-RPS in the cluster, low-weight hosts are allowed to increase their RPS up to their fallback-behavior's limit.  This is limited by the amount of unused-RPS, so as usage increases it will move towards the un-boosted limit (i.e. weight only), and the overall RPS should converge at the target RPS.

Under high usage, it looks more like this (only a small boost added to low weight hosts):
ratelimiter - changed high usage

_How_ the unused RPS is best-distributed is... fuzzy.  Here I've taken the easy and low-data-needed option: just use it all, independently, in each low-weight host.  This is more likely to prevent unfairly-blocked requests, but allows more requests in a multi-host burst (until re-balanced).  
Another option considered (and originally planned) was to distribute it between all *contributing* low-weight hosts, but this needs to either exchange more data or put target-RPS knowledge into both the limiter and the aggregator side of this system.  I think the latter will eventually happen and it's basically fine if/when it does, but it's a larger change than I wanted to do right now, and the former just didn't feel worth the extra logical complexity.  Simpler is better with this kind of system, because it's more understandable and predictable.

The original idea has a simulated implementation here, with a lot of tests to run a simulated traffic pattern and log how weights / individual hosts / etc behave: https://github.com/uber/cadence/pull/6196
but that will likely not be merged - that PR served its purpose of showing whether or not my intuition about the system's behavior as a whole was correct, and the large chunk of code to do that is probably not worth maintaining.

### CODE REVIEW

File: common/quotas/global/collection/collection.go [hunk 1]
```diff
@@ -463,15 +464,18 @@ func (c *Collection) doUpdate(since time.Duration, usage map[shared.GlobalKey]rp
...
 			// so unrecognized values lead to fallback behavior because they cannot be understood.
 			c.global.Load(lkey).FailedUpdate()
 		} else {
-			target := float64(c.targetRPS(lkey))
-			c.global.Load(lkey).Update(rate.Limit(weight * target))
+			target := rate.Limit(c.targetRPS(lkey))
+			limiter := c.global.Load(lkey)
+			fallbackTarget := limiter.FallbackLimit()
+			boosted := boostRPS(target, fallbackTarget, info.Weight, info.UsedRPS)
+			limiter.Update(boosted)
```
Comment (Groxx) [inline common/quotas/global/collection/collection.go:478]: everything else is in service of this change here: boost limits before applying them.

**Review by davidporter-id-au — COMMENTED** [review]
At first glance I can't fault anything here, I'll try and take another look again tomorrow when less tired. 

I appreciate the effort you've gone to explaining visually, mathematically what you're thinking and I'm sure future engineers will similarly. This sets a high bar for us.

**Review by davidporter-id-au — APPROVED** [review]

**Comment by davidporter-id-au** [comment]
I would suggest putting The PR docs probably need to be thrown into a MD file, describing hwo to think about the rate-limiter's tradeoffs and algorithm

File: service/history/handler/handler_test.go [hunk 2]
```diff
@@ -3641,7 +3641,16 @@ func TestRatelimitUpdate(t *testing.T) {
 	w, err := rpc.TestAnyToWeights(t, resp.Any)
 	require.NoError(t, err)
 	assert.Equalf(t,
-		map[shared.GlobalKey]float64{"test:domain-user-limit": 1},
+		map[shared.GlobalKey]rpc.UpdateEntry{
+			"test:domain-user-limit": {
+				Weight: 1,
+				// re 10 vs 15: used RPS only tracks accepted.
```
[the remark this answers is not in the document]
Comment (Groxx) [reply service/history/handler/handler_test.go:3647]:
```go
	update, err := rpc.TestUpdateToAny(t, "testhost", time.Second, map[shared.GlobalKey]rpc.Calls{
		"test:domain-user-limit": {
			Allowed:  10,
			Rejected: 5,
		},
	})
```
^ I inserted 10 allowed and 5 rejected calls before querying for weights

**Comment by Groxx** [comment]
> I would suggest putting The PR docs probably need to be thrown into a MD file, describing hwo to think about the rate-limiter's tradeoffs and algorithm

yea, I think I'll do that once I'm done with the changes and it's fully "on" in production.  better than delaying things at the moment :|

**Pull request merged**